csharp/src/Apache.Arrow/C/CArrowSchema.cs

```diff
@@ -0,0 +1,559 @@
+                Marshal.FreeHGlobal(schema.format);
+                Marshal.FreeHGlobal(schema.name);
+                Marshal.FreeHGlobal(schema.metadata);
+
+                if (schema.n_children > 0)
+                {
+                    for (int i = 0; i < schema.n_children; i++)
+                    {
+                        FreePtr(schema.children[i]);
+                    }
+                    Marshal.FreeHGlobal((IntPtr)schema.children);
+                }
+
+                if (schema.dictionary != IntPtr.Zero)
+                {
+                    FreePtr(schema.dictionary);
+                }
+                Marshal.DestroyStructure<CArrowSchema>(self);
+            };
+
+            private_data = IntPtr.Zero;
+        }
+
+        /// <summary>
+        /// Initialize the exported C schema as a field.
+        /// </summary>
+        /// <param name="field">Field to export.</param>
+        public CArrowSchema(Field field) : this(field.DataType)
+        {
+            name = StringUtil.ToCStringUtf8(field.Name);
+            // TODO: field metadata
+            metadata = IntPtr.Zero;
+            flags = GetFlags(field.DataType, field.IsNullable);
+        }
+
+        /// <summary>
+        /// Initialize the exported C schema as a schema.
+        /// </summary>
+        /// <param name="schema">Schema to export.</param>
+        public CArrowSchema(Schema schema) : this(new StructType(schema.Fields.Values.ToList()))
+        {
+            // TODO: top-level metadata
+        }
+
+        /// <summary>
+        /// Allocate an unmanaged pointer and copy this instances data to it.
+        /// </summary>
+        /// <remarks>
+        /// To avoid a memory leak, you must call <see cref="FreePtr"/> on this
+        /// pointer when done using it.
+        /// </remarks>
+        public IntPtr AllocateAsPtr()
+        {
+            IntPtr ptr = Marshal.AllocHGlobal(Marshal.SizeOf(this));
+            Marshal.StructureToPtr<CArrowSchema>(this, ptr, false);
+            return ptr;
+        }
+
+        /// <summary>
+        /// Free a pointer that was allocated in <see cref="AllocateAsPtr"/>.
+        /// </summary>
+        /// <remarks>
+        /// Do not call this on a pointer that was allocated elsewhere.
+        /// </remarks>
+        public static void FreePtr(IntPtr ptr)
+        {
+            var schema = Marshal.PtrToStructure<CArrowSchema>(ptr);
+            if (schema.release != null)
+            {
+                // Call release if not already called.
+                schema.release(ptr);
+            }
+            Marshal.FreeHGlobal(ptr);
+        }
+
+        /// <summary>
+        /// Export to an existing pointer
+        /// </summary>
+        /// <param name="ptr">An allocated but uninitialized pointer.</param>
+        public void Export(IntPtr ptr)
+        {
+            Marshal.StructureToPtr<CArrowSchema>(this, ptr, false);
+        }
+
+        public static IntPtr AllocateUninitialized()
```

Comment:
This method needs documentation, and should mention needing to call `FreePtr`. The examples in the `Import` methods below should probably also include calls to `FreePtr` for completeness.